### ISSUE TITLE
Fan-in of activities has duplicate TaskScheduledIds when responses are large (serialized to blob storage), causing Orchestrator to hang

### ISSUE THREAD

**Issue by starkna** [Azure/azure-functions-durable-extension#780]
I have a fairly simple fan-out/fan-in orchestrator that starts up between 40-100 activities.  These activities are compute-intensive and require around 1-2 minutes to execute.  I have limited the maxConcurrentActivityFunctions to 1 in the host.json file.  During my testing of the function locally in a development environment, everything worked fine, but when I run in Azure on the consumption plan, I am running into a problem where all activities always complete successfully, but the Orchestrator never gets past my Task.WhenAll statement that is waiting on all of the activity tasks to complete.  No exceptions seem to be thrown.  I have tried running with and without extended sessions, but both run into this same issue.

- Durable Functions extension version:  1.8.1
- Function App version (1.0 or 2.0):  2.0
- Programming language used:  C#

Azure Deployment Details:
- Timeframe issue observed:  Always
- Function App name:  DdsMfgDataGeneration-SL1-Dev
- Function name(s):  MfgDataGenerationOrchestrator
- Region:  CentralUS
- Orchestration instance ID(s):  d5fb96a743204343820aeba9e57b5178

Below is the source code from my Orchestrator function:
```
[FunctionName("MfgDataGenerationOrchestrator")]
        public static async Task RunOrchestrator(
            [OrchestrationTrigger] DurableOrchestrationContext context, ILogger log)
        {
            bool errors = false;
            var request = context.GetInput<MfgDataGenerationWorkRequest>();

            // retry at 10, 30, and 90 seconds before failing
            var retryOptions = new RetryOptions(TimeSpan.FromSeconds(10), 3) { BackoffCoefficient = 2.0 };

            try
            {
                // First begin the work request
                request = await context.CallActivityWithRetryAsync<MfgDataGenerationWorkRequest>("BeginWork", retryOptions, request);

                    var workerRequests = await context.CallActivityAsync<MfgDataGenerationWorkerRequest[]>("GetWorkerRequests", request);

                    // Now create a task for each data generation
                    Task<MfgDataGenerationResult>[] mfgDataTasks = new Task<MfgDataGenerationResult>[workerRequests.Length];
                    for (int i = 0; i < workerRequests.Length; i++)
                    {
                        mfgDataTasks[i] = context.CallActivityWithRetryAsync<MfgDataGenerationResult>("GenerateMfgData", retryOptions, workerRequests[i]);
                    }
             
                    var mfgDataResults = await Task.WhenAll(mfgDataTasks);
                    log.LogInformation("WhenAll completed.");

                        var modelMfgs = new List<ModelMfg>();

                        // Process all of the mfg data results into the model mfgs and save
                        bool error = false;
                        string errorDescription = string.Empty;
                        foreach (var mfgResult in mfgDataResults)
                        {
                            if (mfgResult.Error)
                            {
                                error = true;

                                if (mfgResult?.ErrorDescription != null)
                                    errorDescription += $"{mfgResult.ErrorDescription}{Environment.NewLine}";
                            }
                            else
                            {
                                modelMfgs.Add(mfgResult.ModelMfg);
                            }
                        }

                        request.ModelMfgs = modelMfgs;

                        // Create a save request and send
                        request.UpdateModelSaveParameters(
                            string.IsNullOrEmpty(errorDescription) ? "dds worker mfg data generation complete" : errorDescription,
                            true,  // Do not checkin if more work to do
                            error ? SubmitAction.None : SubmitAction.Submit, // Do not submit if any errors were reported
                            updatedModelMfgs: modelMfgs);
                        request = await context.CallActivityWithRetryAsync<MfgDataGenerationWorkRequest>("CompleteWork", retryOptions, request);
            }
            catch (Exception)
            {
                errors = true;
                throw;
            }
            finally
            {
                // Attempt a save if there were any errors or if a timeout occurred
                if (errors || timeout)
                {
                    request.UpdateModelSaveParameters(
                        timeout ? $"dds worker timeout occurred" : "error occurred during dds worker processing",
                        true, SubmitAction.None);
                    await context.CallActivityWithRetryAsync("CompleteWork", retryOptions, request);
                }
            }
        }
```

**Expected behavior**
Orchestrator should execute past the awaited Task.WhenAll after all activities complete.

**Actual behavior**
Code after the Task.WhenAll is never executed.

**Known workarounds**
None I have been able to find yet.

Development environment: Visual Studio 2019
    "durableTask": {
      "extendedSessionsEnabled": true,
      "extendedSessionIdleTimeoutInSeconds": 60,
      "maxConcurrentActivityFunctions": 1,
      "partitionCount": 1  -- I have tried leaving as default 4 as well
    }

**Comment by starkna** [comment]
This may be a similar problem as reported in issue #774

**Comment by starkna** [comment]
After further digging through the DurableFunctionsHubHistory table, I discovered that there are several duplicate TaskScheduledIds in the 60 TaskCompleted events.  There should have been 60 activities scheduled here, and 60 activities completed, but several had duplicate TaskScheduledIds.  Is this an indication of a race condition on returning the task result to the caller?  Note that all of the results are being stored in blobs due to their large content.

**Comment by starkna** [comment]
Searching for other related issues based on the new information about the TaskScheduledId duplication, this seems to be the same as issue #559 - Is there any estimate on when a fix for this may occur, or if a workaround to that issue exists?  It seems like a serious problem if the activity results are being returned into the wrong task id, if I'm understanding that issue correctly.  Thanks!

**Comment by cgillum** [comment]
We released the fix for the duplication problem that causes orchestrations to get stuck. The fix can be acquired by updating to [v1.8.2](https://github.com/Azure/azure-functions-durable-extension/releases/tag/v1.8.2).

@starkna please give this release a try and see if it resolves the issues you were encountering. I wouldn't expect large messages to be related, but that would certainly be interesting if there is a separate issue related to large messages not covered by this fix.

**Comment by starkna** [comment]
@cgillum I have updated to the 1.8.2 release this morning and re-run the same test.  Unfortunately, I have the same issue with this latest release, so this issue seems to still be unresolved.  My orchestrator schedules and starts 60 activities; I also see 60 completions in the logs, but several of them have duplicate TaskScheduledIds; it seems as if the activities are completing and returning associating their results to the wrong task id.  The end result is that the orchestrator seems to hang, presumably waiting for the task ids that never return.  

However, if I do not return the result directly from my activity and instead write it to a temporary blob storage location, which I then read after all activities complete in a post-processing activity, everything is working fine.

**Comment by mpaul31** [comment]
what if you don’t fan-in and instead await each task result? Same result?

**Comment by starkna** [comment]
Scheduling/awaiting each task in serial works fine.  Or are you suggesting scheduling all at once and awaiting each instead of a Task.WhenAll?

**Comment by cgillum** [comment]
@starkna Is it the same function app as before? Can you share the instance ID for your most recent failed test?

**Comment by starkna** [comment]
@cgillum Yes, same, function app.  The instance ID of this test was 0c60ac06360444639018629ce70dba97.  Let me know if I can do anything else to help narrow down the issue.  Thanks!

**Comment by starkna** [comment]
@cgillum Just wanted to check in if there was any updates on this.  Also, in the last few days, we have run into a new problem that occurs with some of our orchestrations.  In the most recent run, we have scheduled 31 activity tasks, and 12 of them completed successfully, but the others have never even been started from what I can see in the logs.  The function app is DdsMfgDataGeneration-Research, the execution id is b23bc3f94ece428b9e7a855cadeb5add, and the orchestration instance id is c94f33f614544bc78223eae78b0d5ed8.

**Comment by cgillum** [comment]
Hey @starkna, would you mind actually creating a new GitHub issue for these investigations? That will make it easier for me to stay on top of them.

I was able to find the orchestration you mentioned and do see that it appears to have gotten stuck, but need to dig a bit deeper to try and figure out what happened.

**Comment by starkna** [comment]
@cgillum Sorry about that - I have created a new issue report for this problem.  Thanks.
